### ISSUE TITLE
Histogram: class for accumulating binned value counts and outliers

### ISSUE THREAD

**Issue by bukkideme** [ScottPlot/ScottPlot#2389]
The recent histogram method requires an already existing double array. 
I started to think about a special method we could call it Rolling Histogram, or like Point by point Histogram as they call such things in LabVIEW (for example [https://www.ni.com/docs/en-US/bundle/labview/page/ptbypt/prob_and_stat_vis_ptbypt.html,](https://www.ni.com/docs/en-US/bundle/labview/page/ptbypt/prob_and_stat_vis_ptbypt.html) ) where we can Enqueue values, and the histogram bin counts are automatically refreshed at each iterations. We do not want to store the data array (would cause a memory leak on long term run), but only the counts for the bins.

What do you think?

**Comment by bukkideme** [comment]
I think now the most simple approach is just to wait for N samples, for example N=100 values, do the binning, then wait again for a new 100 samples, do the binning, etc. After each binning, we just need to add the new "counts" array to the previous one, so we accumulate the values. With a fixed min/max limit, the bin positions will be fixed, so I think this is the most straightforward way to handle "rolling" statistics...?

**Comment by swharden** [comment]
> Rolling Histogram ... where we can Enqueue values ... and the histogram bin counts are automatically updated

> What do you think?

I love this idea! I wouldn't modify the existing classes, but we could certainly create a `LiveHistogram` class that does this using pre-defined bins.

If the bins are predefined and evenly spaced, the bin counts could be a `readonly int[]`, and adding each new datapoint would bump the value in that count array at the correct index by one.

Is this something you'd consider creating and sharing? You could copy/paste the code of this class in this comments area, or submit a pull request. A cookbook example would be great, but I'm happy to create that if it makes your job easier.

As far as usage, I'm open to suggestions, but I'm envisioning something like:

```cs
int firstBin = 0;
binSize = 5;
binCount = 20;
LiveHistogram hist = new(firstBin, binSize, binCount);

for (double y in GetSensorReadings())
{
    hist.Add(y); // automatically updates counts as new items are added
}

for (int i=0; i<hist.BinCount; i++)
{
    double binLow = hist.Bins[i];
    double binHigh = hist.Bins[i] + hist.BinSize;
    int count = hist.Counts[i];
    Console.WriteLine($"There are {count } values between {binLow} and {binHigh}");
}
```

Then it could be added to a plot and would always display live data like this:
```cs
formsPlot1.Plot.AddScatter(hist.Bins, hist.Counts);
formsPlot1.Refresh(); // call this whenever data is added
```

What do you think?

**Comment by bukkideme** [comment]
Hi! I think it could be like what you showed. I would even add another method, "hist.AddRange(double[] input)", so the feature can be used more flexible.
I put together something based on your example, and I create a pull request. I would also include the new code for the outlier handling in the same pull request if it is fine for you (#2387)? 

Right now I used the following solution in my project (I only update the Graph after every new 10 pcs of data points for performance reasons):

```
//initialization part
// display the histogram counts as a bar plot
private List<double> LatencyGraphSmallList = null;
private double[] LatencyHistArray = null;

HistMinVal = int.Parse(tbLatencyMin.Text);
HistMaxVal = int.Parse(tbLatencyMax.Text);
HistBinSize = double.Parse(tbBinSize.Text);

 (double[] counts, double[] binEdges) =
                        ScottPlot.Statistics.Common.Histogram(new double[] { HistMinVal - 1 }, min: HistMinVal, max: HistMaxVal, HistBinSize);

LatencyHistArray = new double[counts.Length];
LatencyGraphSmallList = new List<double>();
 double[] leftEdges = binEdges.Take(binEdges.Length - 1).ToArray();
plotHistogram.Plot.Clear();
LatencyBarPlot = plotHistogram.Plot.AddBar(values: counts, positions: leftEdges);
LatencyBarPlot.BarWidth = HistBinSize; 
```

```
//In the loop where I get the measurement data:
LatencyGraphSmallList.Add(latency);
 if (latency > HistMaxVal)
{
       AboveMaxCount++;
       tbAboveMax.Text = $"{AboveMaxCount}";
}
if (LatencyGraphSmallList.Count >= 10)
{
       (double[] counts, double[] binEdges) =
       ScottPlot.Statistics.Common.Histogram(LatencyGraphSmallList.ToArray(), min: HistMinVal, max: HistMaxVal, HistBinSize);
       double[] leftEdges = binEdges.Take(binEdges.Length - 1).ToArray();
       LatencyHistArray = LatencyHistArray.Zip(counts, (val1, val2) => (val1 + val2)).ToArray();

       LatencyBarPlot.Replace(leftEdges, LatencyHistArray);
       plotHistogram.Refresh();
       LatencyGraphSmallList.Clear();
}
```

**Comment by swharden** [comment]
Thanks for the additional information!

> hist.AddRange(double[] input)

I like this idea too

> I would also include the new code for the outlier handling (#2387) in the same pull request if it is fine for you? 

I would actually recommend two separate PRs, if that's okay with you. This way requesting changes to one feature won't delay the other.

> Right now I used the following solution in my project (I only update the Graph after every new 10 pcs of data points for performance reasons)

Thanks for sharing this code!

If you start a PR to create this new type of histogram, I can extend your PR to add a demo that simulates live incoming data and redraws the chart after 10 updates (or perhaps every 100 milliseconds?)

**Comment by bukkideme** [comment]
Hi! 

Sure, having separate PRs makes the things easier to handle, ok!
I think the outlier handling PR will be the first, easier to make. I would extend the outlier handling with two new public fields. If the user enables outlier including in border bins, it would be very useful to access the min and max outlier number info, what do you think? These public fields would be also accessible if the "bool includeOutliers" is set to false, if the user does not want to put these counts into the border bins, but use these separately...What do you think?

And it would be great yes, if you can extend my LiveHistrogram PR after accepting it, with a demo. In my case my data rate is 100 msec, and I refreshed the plot at every second, so after every new 10 points.
Thanks!

**Comment by bukkideme** [comment]
Hi!

I am quite a beginner in GIT things, I am not sure what is the proper way to create a second PR, not including my code changes proposed in my first PR. So I have my Fork of the Scottplot/main, now if I want to make a second PR, but not including my code changes made in my first PR, what should I do? Should I make a new branch, or should I checkout/detach "to the commit" before my last changes?
Of course I can just wait until my first PR gets merged into main (if accepted), but it would be nice to know how I can work in parallel in my single Fork, so creating independent PRs based on the last merged commit in main...
Thanks! :)

![image](https://user-images.githubusercontent.com/51174971/219941381-2ce96985-b970-4e8a-84bd-063efb4b4eb1.png)

**Comment by swharden** [comment]
Summarizing https://github.com/ScottPlot/ScottPlot/pull/2401#issuecomment-1436011131:
* There is a `ScottPlot.Statistics.Histogram` class which has been marked obsolete for two years
* I will delete its content then we can re-use this class name to build out the features described here
* In #2401 I'll add some fields I think will be helpful (to communicate my recommended API design), but I won't add any of the logic. Feel free to modify this class any way you think is best! We can refine it in future PRs 🚀

**Comment by swharden** [comment]
Hi @bukkideme, I just merged #2401 which includes the methods you added. However, I think this new `Histogram` class offers a lot of great opportunity for a simpler way to achieve the same goal! It also has the advantage of supporting live data.

Here's the class I started. It doesn't work yet, but I figured you could extend it from here! I'll put this down for now, but if you create a new PR that adds functionality to this class I will be happy to review it whenever it's ready. If you have git questions along the way, I'm happy to help with that too! 

https://github.com/ScottPlot/ScottPlot/blob/9f54032a2cfec6831c15663da0450339660f977a/src/ScottPlot4/ScottPlot/Statistics/Histogram.cs#L9-L78

Thank you for your effort on this! I agree that the existing histogram methods are a bit clunky, and giving users a simpler way to create histograms would be a great addition to this library 🚀

**Comment by bukkideme** [comment]
Hi!
Thanks! :) I will start to work on the Histogram.cs class, and implement the "point by point" histogram feature.
Some comments: 

1. now we will have a Common.Histogram static methods in the Common static class, and also a Histogram.Histogram method in the "normal" Histogram class. If we only implement "LiveHistogram" methods in the Histogram.cs class, maybe it would be wise to rename it to LiveHistogram.cs ?
2. I think AddOutliersToEdgeBins should not be a settable property, because it is not really normal thing to change this bool on the fly during iterations. I would prefer this as a readonly field, which can be set in the constructor of the class.

What do you think?
Thanks!

**Comment by bukkideme** [comment]
Another thing I am not sure how to deal with: it would be nice to give at least 2 options to the user to construct the Histogram class. Either the user can specify min, binWidth and binCount, or can specify min, max and binCount. In the former, the max bin value will be calculated, in the last case the binWidth will be calculated. Both can be a valid user case, depending on the task to program. 
Now the first issue is that, we cannot make two constructors (constructor overload was my first idea) with same parameter signatures :) 
They would be like these: 
```
public Histogram(double min, double binSize, int binCount)

public Histogram(double min, double max, int binCount)
```
Or should we do something like this, but it can be confusing for the user (well, we can explain in details in the summary, so maybe not?):
```
/// <summary>
/// Create a histogram which will count values supplied by <see cref="Add"/> and <see cref="AddRange"/>.
/// If binSize specified, the max input will be ignored and calculated by values min and binSize 
/// </summary>
public Histogram(double min, double max, int binCount, double binSize = NaN)
```

**Comment by swharden** [comment]
> maybe it would be wise to rename it to LiveHistogram.cs

I considered this, but the class could be used for non-live histogram data to (e.g., users who want access to that outlier information) so I think the more general purpose name makes sense here

> now we will have a Common.Histogram static methods in the Common static class, and also a Histogram.Histogram method in the "normal" Histogram class

True! I think they're different enough use cases that they won't be too confusing though:

```cs
// this is what's there now, but it's not very easy to work with
(double[] hist, double[] binEdges, int minOutliers, int maxOutliers) = ScottPlot.Statistics.Common.Histogram(values, min, max, binSize)
Console.WriteLine(minOutliers);
```

```cs
// this could do the same thing after this work is complete
var histogram = new ScottPlot.Statistics.Histogram(firstBin: 0, binSize: 5, binCount: 20);
histogram.AddRange(values);
Console.WriteLine(histogram.MinOutlierCount);
```

> AddOutliersToEdgeBins ... readonly field

Moving this to the constructor makes sense if we are confident we will not add additional configuration options in the future. Something I'd like to avoid is making the constructor longer and longer as new features and options are added. I'm happy to go either way on this one though, so whatever you think is best I'm fine with.

> specify min, max and binCount

I guess there are two options for how we could define bins...

Define bins by focusing on bin size:
* first bin: 0
* bin size: 5
* bin count: 20

Define bins by focusing on the total range:
* min: 0
* max: 100
* count: 20

Glancing at numpy, it looks like it does it the second way, so I'm in favor of that

https://numpy.org/doc/stable/reference/generated/numpy.histogram.html

```cs
public Histogram(double min, double max, int binCount) { }
```

>  factory method technique for the above 2 variations?

True, but I like the simplicity of having one "right" way to do things. I think the min/max/count is the way to go here!

**Comment by bukkideme** [comment]
Hi!

Thanks for the feedback! 

_True! I think they're different enough use cases that they won't be too confusing though:_
Good! So you mean, we work toward making the ScottPlot.Statistics.Histogram class better and covering all Histogram cases, and we keep now the existing static ScottPlot.Statistics.Common.Histogram methods as temporary options?

_I think the min/max/count is the way to go here!_
This is what I also prefer better, and the resulted binSize can be read as a private settable property...So lets keep only this single option, and a single constructor is enough!

I will go on according to the above considerations, thanks!
